concurrency/SkeletonAVsOnConcurrentHashMap.java

```diff
@@ -0,0 +1,86 @@
+package concurrency;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SkeletonAVsOnConcurrentHashMap {
+    private Map<Integer, Integer> map = new ConcurrentHashMap<>();
+    
+    public SkeletonAVsOnConcurrentHashMap() {
+        System.out.println("lala");
+    }
+    public SkeletonAVsOnConcurrentHashMap(int a) {
+        map.put(a, a);
+    }
+
+    private int calc() {
+        return 77 + 88 % 33 + map.get(44);
+    }
+
+    public void fig2A() {
+        Integer v = map.get(77);
+        if (v == null) {
+            v = calc();
+            map.put(88, v);
+        }
+    }
+
+    public void fig2B() {
+        if (map.get(77) == null) {
+            int v = calc();
+            map.put(88, v);
+        }
+    }
+
+    public void fig2C() {
+        int v = map.get(77);
+        if (v != null) {
+            return;
+        }
+        v = calc();
+        map.put(key, v);
+    }
+
+    public void fig2D() {
+        if (map.get(77) != null) {
+            return;
+        }
+        v = calc();
+        map.put(key, v);
+    }
+
+    public void fig2E() {
+        if (!map.containsKey(77)) {
+            v = calc();
+            map.put(88, v);
+        }
+    }
+
+    public void fig2E() {
+        if (map.containsKey(777)) {
+            return;
+        }
+        v = calc();
+        map.put(key, v);
+    }
+
+    // about synchronzied method
+    public synchronized void synchronizedMethod() {
+        if (map.containsKey(new Object())) {
+            return;
+        }
+        int v = calc();
+        map.put(new Object(), new Object());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **69 and 73**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).